frontend/packages/ceph-storage-plugin/src/selectors/index.ts

```diff
@@ -30,6 +19,27 @@ export const getCephPVs = (pvsData: K8sResourceKind[] = []): K8sResourceKind[] =
     );
   });
 
+export const getCephPVCs = (
+  cephSCNames: string[] = [],
+  pvcsData: K8sResourceKind[] = [],
+  pvsData: K8sResourceKind[] = [],
+): K8sResourceKind[] => {
+  const cephPVs = getCephPVs(pvsData);
+  const boundPVCs = pvcsData.filter((pvc) => _.get(pvc, 'status.phase') === 'Bound');
+  const unboundPVCs = pvcsData.filter((pvc) => _.get(pvc, 'status.phase') !== 'Bound');
+  const filteredBoundPVCs = boundPVCs.filter((pvc) =>
+    cephPVs.some((pv) => _.get(pv, 'spec.claimRef.uid') === _.get(pvc, 'metadata.uid')),
+  );
+  const cephSCsFromPVs = cephPVs.map((pv) => _.get(pv, 'spec.storageClassName'));
+  const filteredUnboundPVCs = unboundPVCs.filter(
+    (pvc) =>
+      cephSCNames.includes(_.get(pvc, 'spec.storageClassName')) ||
+      _.get(pvc, ['metadata', 'annotations', 'volume.beta.kubernetes.io/storage-class']) ||
+      cephSCsFromPVs.some((scName) => scName.includes(_.get(pvc, 'spec.storageClassName'))),
+  );
+  return [...filteredBoundPVCs, ...filteredUnboundPVCs];
```

Comment:
Some comments would go a long way here. I think this is easier to understand as one filter loop and would perform better as well. There could be a lot of PVCs in the cluster, and this will be reevaluated whenever any storage class, PVC, or PV changes. This could very easily be a hot spot in our code.

I'd try something like (untested)

```ts
const getPVCStorageClass = (pvc: K8sResourceKind): string => {
  return pvc.spec.storageClassName || _.get(pvc, ['metadata', 'annotations', 'volume.beta.kubernetes.io/storage-class']);
};

const getPVStorageClass = (pv: K8sResourceKind): string => {
  return pv.spec.storageClassName;
};

export const getCephPVCs = (
  cephSCNames: string[] = [],
  pvcData: K8sResourceKind[] = [],
  pvData: K8sResourceKind[] = [],
): K8sResourceKind[] => {
  const cephPVs = getCephPVs(pvData);
  const cephSCNameSet = new Set<string>([...cephSCNames, ...cephPVs.map(getPVStorageClass)];
  const boundCephPVCUIDs = new Set<string>(_.map(cephPVs, 'spec.claimRef.uid'));
  return pvcData.filter((pvc: K8sResourceKind) => {
    if (pvc.status && pvc.status.bound) {
      return boundCephPVCUIDs.has(pvc.metadata.uid);
    }
    const storageClassName = getStorageClassName(pvc);
    return cephSCNameSet.has(storageClassName);
  });
};
```

Reply:
I have made changes according to your suggestions. However, calling `getCephPVs` is going to kill out all the gain from here since it uses a nested loop.

Reply:
> I have made changes according to your suggestions. However, calling `getCephPVs` is going to kill out all the gain from here since it uses a nested loop.

It's not a problem there since the `getCephPVs` inner loop is iterating over a fixed list of 3 items. That helper is still linear time complexity.

Reply:
My bad I missed out on the detail.